LSPS4/README.md

```diff
@@ -0,0 +1,1788 @@
+
+### 1.  Request SCID
+
+The client can request two types of SCIDs:
+
+* Permanent.
+  The client can have at most one permanent SCID.
+* Ephemeral.
+  The client can have any number of ephemeral SCIDs, up to some LSP-defined
+  limit.
+
+A permanent SCID is actually only permanent as long as the LSP and client
+have a channel in "normal operation" state (after both peers have sent
+`channel_ready`, and before either has sent `shutdown`, and no commitment
+transaction has confirmed).
+If the LSP thinks there are no more "normal operation" channels, the
+permanent SCID has a defined "expiry time"; if the current time has
+exceeded the expiry time then the LSP forgets the permanent SCID.
+
+An ephemeral SCID also has an "expiry time", and is forgotten by the
+LSP as soon as the expiry time has passed, regardless of the existence
+or non-existence of a channel between the LSP and the client.
+It is also forgotten if at least one forwarded payment using that SCID
+has been successfully claimed (`update_fulfill_htlc`) by the client
+(but not if the client fails the HTLC).
+
+* Permanent: Forget if past expiry time AND no channel.
+* Ephemeral: Forget if past expiry time OR payment succeeded.
+
+> **Rationale** All SCIDs, including the "permanent" one, have an
+> expiry time in order to time-bound any extra information the
+> LSP has to store for clients that request SCIDs (which are shared
+> among all clients of the same LSP) and then never actually use
+> them.
+> The permanent SCID needs an expiry time so that the client has
+> an opportunity to have a channel opened (either by opening one
+> itself, or otherwise purchasing a channel from the LSP via this
+> LSPS or via [LSPS1][] or [LSPS2][]) during this expiry time
+> period.
+
+[LSPS1]: ../LSPS1/README.md
+[LSPS2]: ../LSPS2/README.md
+
+Ephemeral SCIDs are provided to give privacy to clients that are
+willing to use ephemeral node IDs to prevent invoices issued by
+the same client from being correlated.
+For each invoice, the client requests an ephemeral SCID and generates
+a fresh keypair as the node ID; separate invoices thus have a
+different final node ID and final hop SCID.
+On receiving a payment, the client iterates over its known invoice
+keypairs until it is able to decode the onion, then can claim the
+funds.
+
+> **Rationale** If ephemeral SCIDs did not exist, then even if the
+> client used ephemeral node IDs, its invoices could be correlated
+> by looking at the LSP and the SCID from the LSP to the client
+> in the route hints of the invoices.
+> Similarly, if the client did not use ephemeral node IDs, its invoices
+> could be correlated by looking at the payee node ID of the invoices.
+> Thus, privacy requires that a client both implement ephemeral node
+> IDs, *and* request ephemeral SCIDs from the LSP.
+
+Permanent SCIDs are provided to give a stable path to clients.
+For example, the upcoming BOLT12 offers will need a stable permanent
+SCID to be embedded in a blinded route.
```

Comment:
This is a flaw of Route Blinding IMO. Route Blinding [encodes an encrypted path into the invoice](https://github.com/lightning/bolts/blob/master/proposals/route-blinding.md?plain=1#L194) based on SCIDs. SCIDs lifespan is potentially very short, especially in case of mobile wallets with a single channel that gets spliced often.

This flaw should be fixed in the Route Blinding spec, not here. It should encode blinded node ids into these paths. Their lifetime is substantially longer.

Reply:
Node IDs are large (33 bytes, 32 if you X-only it).  The BOLT spec specifically uses SCIDs because they are smaller (8 bytes).  We count bytes here because we expect multiple hops in the `update_add_htlc` and the onion size is limited to 1366 bytes total (and that already has to include multiple hops). We also need 32 bytes for MAC per hop already, so using SCIDs cuts down per-hop size quite a bit.

(We should note that route blinding is basically just a copy-paste of the BOLT4 onion encoding; the only substantive change is that in route blinding, the latter part of the BOLT4 onion is encoded using an ephemeral key from the receiver while the earlier part is encoded using an ephemeral key from the sender, whereas the BOLT4 of today has it entirely encoded using an ephemeral key from the sender.  If you want to use node IDs instead of SCIDs here, you need to modify the BOLT4 spec as well, not just the route blinding spec )

Reply:
FINALLY ephemeral SCIDs and ephemeral node IDs in LSPS4 are for non-route-blinded BOLT11 use.  The ephemeral SCIDs and ephemeral node ID prevent the payer from learning the identity of the payee using the BOLT11 invoice (the payee might leak their privacy another way, but at least it will not happen by the BOLT11 invoice).  The resulting BOLT11 invoice is payable using standard payers today that have not even implemented route blinding yet.

Reply:
They should have just used the last 8 bytes of the node id to indicate the next hop if they want to save space 🤷‍♂️ Anyway, I see the complexity of such a change which is impracticle.

What I am trying to do here is to reduce the complexity of the spec by challenging its assumptions.
With what you are saying, I think of the following question:
- Why do we add a privacy feature to BOLT11 if BOLT11 will be replaced by BOLT12 in the near future anyway?

Maybe we just need 1 permanent SCID to solve the "SCID are changing all the time" issue. This would also give us the ability to create invoices offline too. No ephemeral SCIDs needed therefore.

Reply:
>Why do we add a privacy feature to BOLT11 if BOLT11 will be replaced by BOLT12 in the near future anyway?

As you have pointed out, these things take time to implement and it is likely that a majority of implemented services will service BOLT11 (because it is already widely supported today) and the uptake of BOLT12 is likely to lag for a few more years.

Reply:
As I mentioned in the online meeting:

* The LSP-side API that gives the ephemeral SCIDs is going to pretty much call the same SCID generation function, and ephemeral and permanent and "actual" SCIDs are all in the same namespace and you need to ensure they do not conflict anyway.
* Clients do not need to immediately implement the multiple-node-IDs thing.
* Clients that DO implement the multiple-node-IDs thing can issue an invoice that is 100% seamlessly compatible with existing BOLT11 payers, so even if custodial services remain stuck with BOLT11 and do not support BOLT12, they can still pay to privacy-enhanced BOLT11 invoices.
  * Custodial services not implementing this are fine since they are unlikely to need JIT channels at all, they are likely to have pre-arranged inbound liquidity (e.g. LSPS1) and individual users of the custodial service already get some privacy anyway since all invoices they generate point at the custodial service, not their own individual Lightning node.

Reply:
Ephemeral SCIDs have been removed.